### PR TITLE
[RIPD-772] Validator key manifests (take 2)

### PR DESCRIPTION
**Pull request by josh-ripple**
* Fix some things
* Support ephemeral validator keys in UniqueNodeList
* Add new ManifestCache
* Trade validator key manifests via peer gossip

This pull request supersedes <https://github.com/ripple/rippled/pull/978>.

Reviewers: @scottschurr @ximinez

### CODE REVIEW

File: src/ripple/overlay/impl/OverlayImpl.cpp [hunk 1]
```diff
@@ -570,10 +609,62 @@ OverlayImpl::onPeerDeactivate (Peer::id_t id,
...
+    for_each( [&](std::shared_ptr<PeerImp> const& peer)
+        {
+            if (peer != from)
+            {
+                if (auto const& j = peer->pjournal().debug) j
+                    << "... to " << peer.get();
+                peer->send(msg);
+            }
+        });
+}
```
Comment (JoelKatz) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: This is inferior to the code we use elsewhere. For example, if a manifest is received from three peers in the time it takes us to sign it, we'll relay it to every peer but the one we received it from first. We also can only signature check one manifest at a time, regardless of which peer(s) we received it from. If we receive a lot of manifests from a lot of peers, all our threads will wind up blocked on the mutex trying to check manifest signatures.

Comment (vinniefalco) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: Also, we need to charge the peer's resource consumption endpoint if the signature verification fails.

Comment (josh-ripple) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: I modified `applyManifest()` so it releases the lock before verifying the signature, then re-acquires it and checks the prerequisite conditions again.

**Comment by ximinez** [comment]
There are some issues brought up in other comments, but I don't have anything else to add, so I'll give a tentative :+1: and follow any additional commits before final sign off.

**Comment by scottschurr** [comment]
@josh-ripple: If you're punting on `VersionEncoding`, then I think that's the last of my concerns.  I think it's good to go assuming that @JoelKatz and @vinniefalco are happy with the `Don't verify signatures under mutex lock` fix.  So I'll give you a tentative...
:+1:

**Comment by vinniefalco** [comment]
This code has to go back to the drawing board at least somewhat, to change the way it queues signature checks. And it should use the HashRouter to squelch pointless forwards.

**Comment by ximinez** [comment]
Travis failure on 94861ee appears legit. The changes look good so far, but the build fails in MSVC, too.
```1>..\..\src\ripple\overlay\tests\manifest_test.cpp(56): error C2885: 'accepted': not a valid using-declaration at non-class scope
1>..\..\src\ripple\overlay\tests\manifest_test.cpp(57): error C2885: 'malformed': not a valid using-declaration at non-class scope
1>..\..\src\ripple\overlay\tests\manifest_test.cpp(58): error C2885: 'untrusted': not a valid using-declaration at non-class scope
1>..\..\src\ripple\overlay\tests\manifest_test.cpp(59): error C2885: 'stale': not a valid using-declaration at non-class scope
1>..\..\src\ripple\overlay\tests\manifest_test.cpp(60): error C2885: 'invalid': not a valid using-declaration at non-class scope
```

**Comment by rec** [comment]
Ping!

File: src/ripple/overlay/tests/manifest_test.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    void
+    run() override
+    {
+        using ManifestDisposition::accepted;
+        using ManifestDisposition::malformed;
+        using ManifestDisposition::untrusted;
+        using ManifestDisposition::stale;
+        using ManifestDisposition::invalid;
```
Comment (ximinez) [inline src/ripple/overlay/tests/manifest_test.cpp]: MSVC doesn't like this syntax / shortcut. I don't know if there's a workaround beyond replacing the abbreviated enums with the fully qualified name.

**Comment by josh-ripple** [comment]
The commits added above are the result of rebasing over `develop` (nixing the addition of `for_each` to Overlay in favor of the one that's now there) and adding three new commits to fix two breakages in the test code and add a new test case.

**Comment by scottschurr** [comment]
@josh-ripple Is it time to give this pull request another review?  Or are you still working on it?

**Comment by josh-ripple** [comment]
The four new commits above are ready.  I'm still working on suppressing duplicate manifest deliveries.

**Comment by scottschurr** [comment]
So you'll be adding more commits before it's ready for a final review?

**Comment by josh-ripple** [comment]
Yes.

**Comment by vinniefalco** [comment]
Closing this until its ready for review.

**Pull request closed without merging**